docs/design/2019-11-14-tidb-builtin-diagnostics-zh_CN.md

```diff
@@ -0,0 +1,715 @@
+# RFC: TiDB 内置 SQL 诊断
+
+## Summary
+
+目前 TiDB 获取诊断信息主要依赖外部工具（perf/iosnoop/iotop/vmstat/sar/...）、监控系统（Prometheus/Grafana）、日志文件、HTTP API 和 TiDB 提供的系统表。分散的工具链和繁杂的获取方式导致 TiDB 的集群的使用门槛高、运维难度大、不能提前发现问题以及遇到问题不能及时排查、诊断和恢复集群等。
+
+本提案提出一种新的方法，在 TiDB 中内置获取诊断信息的功能，并将诊断信息使用系统表的形式对外暴露，使用户可以使用 SQL 的方式进行查询。
+
+## Motivation
+
+本提案主要解决 TiDB 在获取诊断信息过程中的以下问题：
+
+- 工具链分散，需要在不同工具之间来回切换，且部分 Linux 发行版未内置相应工具或内置工具的版本不一致。
+- 信息获取方式不一致，比如有 SQL、HTTP、导出监控、登录各个节点查看日志等。
+- TiDB 集群组件较多，对不同组件的监控进行对比和关联低效且繁琐。
+- TiDB 没有集中日志管理组件，没有高效的手段对整个集群的日志进行过滤、检索、分析、聚合。
+- 系统表只包含当前节点信息，不能体现整个集群的状态，如：SLOW_QUERY, PROCESSLIST, STATEMENTS_SUMMARY。
+
+在通过提供多维度集群级别系统表和集群诊断规则框架之后，提高全集群信息查询、状态获取、日志检索、一键巡检、故障诊断几个使用场景中的效率，并为后续异常预警功能提供基础数据。
+
+## Detailed Design
+
+### 系统整体概览
+
+本提案的实现分为四层：
+
+- L1: 最底层在各个节点实现信息采集模块，包括 TiDB/TiKV/PD 监控信息、硬件信息、内核中记录的网络 IO、磁盘 IO 信息、CPU 使用率、内存使用率等。
+- L2: 第二层通过调用底层信息采集模块并通过对外服务接口(HTTP API/gRPC Service)向上层提供数据，使 TiDB 可以获取当前节点采集到的信息。
+- L3: 第三层由 TiDB 拉取各个节点的信息进行聚合和汇总，并以系统表的形式对上层提供数据。
+- L4: 第四层实现诊断框架，诊断框架通过查询系统表获取整个集群的状态，并根据诊断规则得到诊断结果。
+
+如下从信息采集到使用使用诊断规则对采集的信息进行分析的数据流向图:
+
+```
++-L1--------------+             +-L3-----+
+| +-------------+ |             |        |
+| |   Metrics   | |             |        |
+| +-------------+ |             |        |
+| +-------------+ |             |        |
+| |   Disk IO   | +---L2:gRPC-->+        |
+| +-------------+ |             |        |
+| +-------------+ |             |  TiDB  |
+| |  Network IO | |             |        |
+| +-------------+ |             |        |
+| +-------------+ |             |        |
+| |   Hardware  | +---L2:HTTP-->+        |
+| +-------------+ |             |        |
+| +-------------+ |             |        |
+| | System Info | |             |        |
+| +-------------+ |             |        |
++-----------------+             +---+----+
+                                    | 
+                   +---infoschema---+ 
+                   |                  
+                   v                  
++-L4---------------+---------------------+
+|                                        |
+|          Diagnosis Framework           |
+|                                        |
+| +---------+ +---------+  +---------+   |
+| | rule1   | |  rule2  |  |  rule3  |   |
+| +---------+ +---------+  +---------+   |
++----------------------------------------+
+```
+
+### 系统信息收集
+
+TiDB/TiKV/PD 三个组件都需要实现系统信息采集模块，其中 TiDB/PD 使用 Golang 实现并复用逻辑，TiKV 需要使用 Rust 单独实现。
+
+#### 节点硬件信息
+
+各个节点需要获取的硬件信息包括：
+
+- CPU 信息：物理核心数、逻辑核心数量、NUMA 信息、CPU 频率、CPU 供应商、L1/L2/L3 缓存大小
+- 网卡信息：网卡设备名、网卡是否启用、生产厂商、型号、带宽、驱动版本、接口队列数（可选）
+- 磁盘信息：磁盘名、磁盘容量、磁盘使用量、磁盘分区、挂载信息
+- USB 设备列表
```

Comment:
Why scrape this?

Reply:
In some extreme cases, USB devices may cause system anomalies. (it's not an optional system information)